src/Splat.Microsoft.Extensions.DependencyInjection/MicrosoftDependencyResolver.cs

```diff
@@ -208,34 +205,31 @@ public virtual void UnregisterAll(Type? serviceType, string? contract = null)
 
         lock (_syncLock)
         {
-            switch (contract)
+            if (_serviceCollection is null)
             {
-                case null when _serviceCollection is not null:
-                    {
-                        var sds = _serviceCollection
-                            .Where(s => s.ServiceType == serviceType)
-                            .ToList();
-
-                        foreach (var sd in sds)
-                        {
-                            _serviceCollection.Remove(sd);
-                        }
-
-                        break;
-                    }
-
-                case null:
-                    throw new ArgumentException("There must be a valid contract if there is no service collection.", nameof(contract));
-                default:
-                    {
-                        var dic = GetContractDictionary(serviceType, false);
-                        if (dic?.TryRemoveContract(contract) == true && dic.IsEmpty)
-                        {
-                            RemoveContractService(serviceType);
-                        }
-
-                        break;
-                    }
+                // required so that it gets rebuilt if not injected externally.
+                _serviceProvider = null;
+                return;
+            }
+
+            IEnumerable<ServiceDescriptor> sds = Enumerable.Empty<ServiceDescriptor>();
+
+            if (contract is null || string.IsNullOrWhiteSpace(contract))
+            {
+                sds = _serviceCollection.Where(s => !s.IsKeyedService && s.ServiceType == serviceType);
+            }
+            else
+            {
+                sds = _serviceCollection
+                  .Where(sd => sd.IsKeyedService
+                            && sd.ServiceKey is string serviceKey
+                            && serviceKey == contract
+                            && sd.ServiceType == serviceType);
```

Comment:
putting "&& sd.ServiceType == serviceType" above the "&& sd.ServiceKey is string serviceKey" will short cicruit before the cast and string equality check.